File: Vb/ImageProcessing/ComImgProc.vb
Imports System.Threading

Namespace ImageProcessing
    Public MustInherit Class ComImgProc
        Private m_bitmap As BitmapImage
        Private m_wBitmap As WriteableBitmap

        Public Sub New(ByVal _bitmap As BitmapImage)
            m_bitmap = _bitmap
        End Sub

        Public Property Bitmap() As BitmapImage
            Get
                Return m_bitmap
            End Get
            Set(ByVal value As BitmapImage)
                m_bitmap = value
            End Set
        End Property

        Public Property WriteableBitmap() As WriteableBitmap
            Get
                Return m_wBitmap
            End Get
            Set(ByVal value As WriteableBitmap)
                m_wBitmap = value
            End Set
        End Property

        Public MustOverride Function GoImgProc(ByVal _token As CancellationToken) As Boolean
    End Class
End Namespace

File: CSharp/ImageProcessing/Lib/ComImgProc.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading;
using System.Threading.Tasks;
using System.Windows.Media.Imaging;

/// <summary>
/// 画像処理の共通のロジック
/// </summary>
abstract public class ComImgProc
{
    protected BitmapImage m_bitmap;
    protected WriteableBitmap m_wBitmap;

    /// <summary>
    /// コンストラクタ
    /// </summary>
    /// <param name="_bitmap">ビットマップ</param>
    public ComImgProc(BitmapImage _bitmap)
    {
        m_bitmap = _bitmap;
    }

    /// <summary>
    /// デスクトラクタ
    /// </summary>
    ~ComImgProc()
    {
        m_bitmap = null;
        m_wBitmap = null;
    }

    /// <summary>
    /// 初期化
    /// </summary>
    virtual public void Init()
    {
        m_bitmap = null;
    }

    /// <summary>
    /// ビットマップ
    /// </summary>
    public BitmapImage Bitmap
    {
        set { m_bitmap = value; }
        get { return m_bitmap; }
    }

    /// <summary>
    /// Writeableなビットマップ
    /// </summary>
    public WriteableBitmap WriteableBitmap
    {
        set { m_wBitmap = value; }
        get { return m_wBitmap; }
    }

    /// <summary>
    /// 画像処理実行の抽象
    /// </summary>
    /// <param name="_token">キャンセルトークン</param>
    /// <returns>実行結果 成功/失敗</returns>
    abstract public bool GoImgProc(CancellationToken _token);
}